plutus-ledger/src/Ledger/Constraints/OffChain.hs

```diff
@@ -396,22 +392,22 @@ processConstraint = \case
         unbalancedTx . tx . Tx.forgeScripts %= Set.insert monetaryPolicyScript
         unbalancedTx . tx . Tx.forge <>= value
         -- If i is negative we are burning tokens. This counts as an output, so we should subtract
-        -- the amount burned from valueSpentRequired, just as in `MustPayToPubKey` below.
+        -- the amount burned from valueSpentBalance, just as in `MustPayToPubKey` below.
         -- Subtracting the amount burned is the same as adding the (negative) amount forged.
-        if i < 0 then valueSpentRequired <>= value
-                 else valueSpentActual   <>= value
+        if i < 0 then valueSpentBalance <>= value
+                 else valueSpentBalance  <>= N.negate value
```

Comment:
This seems wrong to me... shouldn't it just be added to the balance? As it is, both `i = 5` and `i = -5` have the same effect on the balance, which surely isn't right.

Reply:
It's `valueSpentBalance = valueSpentRequired - valueSpentActual`, so where we used to add something to `valueSpentActual` we now have to subtract it from `valueSpentBalance`.

Reply:
Then I don't understand how what was there was right either. I still don't see how `i = 5` and `i = -5` should have the same effect on the balance.

Reply:
The balance tracks value that needs to be involved in the transaction (as required by `MustSpendValue`). Both forging and burning tokens counts as being involved. Similarly, both adding an input and adding an output subtracts from the balance.